runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,339 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+            newCallNode->setAndIncChild(i, child);
+            }
+         }
+      }
+   if(callNode->isPreparedForDirectJNI())
+      {
+         newCallNode->setPreparedForDirectJNI();
+      }
+
+   return newCallNode;
+   }
+
 ///////////////////////////////////////////////////////////////////////////////////////
 // multianewArrayEvaluator:  multi-dimensional new array of objects
 ///////////////////////////////////////////////////////////////////////////////////////
 TR::Register *
 J9::Z::TreeEvaluator::multianewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    {
-   TR::ILOpCodes opCode = node->getOpCodeValue();
-   TR::Node::recreate(node, TR::acall);
-   TR::Register * targetRegister = directCallEvaluator(node, cg);
-   TR::Node::recreate(node, opCode);
-   return targetRegister;
-   }
+   static char *useDirectHelperCall = feGetEnv("TR_MultiANewArrayEvaluatorUseDirectCall");
+   TR::Compilation *comp = cg->comp();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(comp->fe());
+   TR::Register *targetReg = cg->allocateRegister();
+
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Node *thirdChild = node->getThirdChild();
+
+   if (useDirectHelperCall || !secondChild->getOpCode().isLoadConst() || secondChild->getInt()!=2)
+      {
+      traceMsg(cg->comp(), "DCDCDCDC - multianewArrayEvaluator. More than 2 dimensions, so we can't handle this scenario in JITT'ed code\n");
+      return performHelperCall(node, NULL, TR::acall, false, cg);
+      }
+
+   TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg);
+   TR::LabelSymbol *nonZeroFirstDimLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *cFlowRegionEnd = generateLabelSymbol(cg);
+   TR::LabelSymbol *oolFailLabel = generateLabelSymbol(cg);
+
+   cFlowRegionStart->setStartInternalControlFlow();
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionStart);
+
+   TR::Register *dimsPtrReg = cg->evaluate(firstChild);
+   TR::Register *dimReg = cg->evaluate(secondChild);
+   TR::Register *classReg = cg->evaluate(thirdChild);
+
+   // First generate the OOL code. This path will be used to handle the scenarios that we don't want to handle in JITT'ed code.
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(oolFailLabel, cFlowRegionEnd, cg);
+   cg->getS390OutOfLineCodeSectionList().push_front(outlinedSlowPath);
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, oolFailLabel);
+
+   TR::Node *_callNode = createOutlinedCallNode(cg, node, TR::acall);
+   TR::Register *targetReg2 = NULL;
+   targetReg2 = TR::TreeEvaluator::performCall(_callNode, false, cg);
+   cg->decReferenceCount(_callNode);
+   generateRRInstruction(cg, TR::InstOpCode::LR, node, targetReg, targetReg2);
```

Comment:
This looks incorrect. What about 64-bit non-compressedrefs? Surely this will fail every time right? Does any of the testing catch this? If it doesn't I'm a bit worried we don't have adequate testing in this area.

Reply:
This should no longer be an issue: ff6da2b